File: apps/android/app/src/main/java/org/llamenos/hotline/ui/notes/NotesViewModel.kt
package org.llamenos.hotline.ui.notes

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.filterNotNull
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonPrimitive
import org.llamenos.hotline.api.ApiService
import org.llamenos.hotline.api.SessionState
import org.llamenos.hotline.crypto.CryptoService
import org.llamenos.hotline.hub.ActiveHubState
import org.llamenos.hotline.model.CustomFieldDefinition
import org.llamenos.hotline.model.NotePayload
import org.llamenos.hotline.model.NoteRepliesResponse
import org.llamenos.hotline.model.NoteReply
import org.llamenos.hotline.model.NotesListResponse
import org.llamenos.protocol.CreateNoteBody
import org.llamenos.protocol.CreateNoteBodyAdminEnvelope
import org.llamenos.protocol.CreateNoteBodyAuthorEnvelope
import org.llamenos.protocol.CreateReplyBody
import org.llamenos.protocol.CreateReplyBodyReaderEnvelope
import org.llamenos.protocol.NoteResponse
import org.llamenos.protocol.RecipientEnvelope
import javax.inject.Inject

/**
 * Decrypted note for UI display. Combines the API response metadata with
 * the decrypted plaintext content.
 */
data class DecryptedNote(
    val id: String,
    val text: String,
    val fields: Map<String, JsonElement>?,
    val authorPubkey: String,
    val callId: String?,
    val conversationId: String?,
    val replyCount: Int = 0,
    val createdAt: String,
    val updatedAt: String?,
)

/**
 * A decrypted note reply for display.
 */
data class DecryptedReply(
    val id: String,
    val authorPubkey: String,
    val text: String,
    val createdAt: String,
)

data class NotesUiState(
    val notes: List<DecryptedNote> = emptyList(),
    val isLoading: Boolean = false,
    val isRefreshing: Boolean = false,
    val error: String? = null,
    val currentPage: Int = 1,
    val totalNotes: Int = 0,
    val hasMorePages: Boolean = false,

    // Search
    val searchQuery: String = "",

    // Note creation / editing
    val customFields: List<CustomFieldDefinition> = emptyList(),
    val isSaving: Boolean = false,
    val saveError: String? = null,
    val saveSuccess: Boolean = false,

    // Note detail
    val selectedNote: DecryptedNote? = null,
    val isEditing: Boolean = false,

    // Thread replies
    val replies: List<DecryptedReply> = emptyList(),
    val isLoadingReplies: Boolean = false,
    val isSendingReply: Boolean = false,
)

/**
 * ViewModel for the Notes feature.
 *
 * Handles fetching, decrypting, and creating encrypted notes.
 * Each note is encrypted with per-note forward secrecy — a unique random key
 * per note, ECIES-wrapped for the author and each admin.
 *
 * Decryption: find the envelope matching our pubkey, unwrap the symmetric key
 * via ECIES, then decrypt the note ciphertext with XChaCha20-Poly1305.
 */
@HiltViewModel
class NotesViewModel @Inject constructor(
    private val apiService: ApiService,
    private val cryptoService: CryptoService,
    private val sessionState: SessionState,
    private val activeHubState: ActiveHubState,
) : ViewModel() {

    private val json = Json { ignoreUnknownKeys = true }

    private val _uiState = MutableStateFlow(NotesUiState())
    val uiState: StateFlow<NotesUiState> = _uiState.asStateFlow()

    init {
        activeHubState.activeHubId
            .filterNotNull()
            .onEach { loadNotes() }
            .launchIn(viewModelScope)
    }

    /**
     * Load notes from the API and decrypt them locally.
     */
    fun loadNotes(page: Int = 1) {
        viewModelScope.launch {
            _uiState.update {
                it.copy(
                    isLoading = page == 1 && it.notes.isEmpty(),
                    isRefreshing = page == 1 && it.notes.isNotEmpty(),
                    error = null,
                )
            }

            try {
                val response = apiService.request<NotesListResponse>(
                    "GET",
                    apiService.hp("/api/notes") + "?page=$page&limit=20",
                )

                val decrypted = response.notes.mapNotNull { note ->
                    decryptNote(note)
                }

                _uiState.update {
                    val allNotes = if (page == 1) decrypted else it.notes + decrypted
                    it.copy(
                        notes = allNotes,
                        isLoading = false,
                        isRefreshing = false,
                        currentPage = page,
                        totalNotes = response.total,
                        hasMorePages = allNotes.size < response.total,
                    )
                }
            } catch (e: Exception) {
                _uiState.update {
                    it.copy(
                        isLoading = false,
                        isRefreshing = false,
                        error = e.message ?: "Failed to load notes",
                    )
                }
            }
        }
    }

    /**
     * Pull-to-refresh — reload from page 1.
     */
    fun refresh() {
        loadNotes(page = 1)
    }

    /**
     * Load the next page of notes.
     */
    fun loadNextPage() {
        val state = _uiState.value
        if (!state.hasMorePages || state.isLoading) return
        loadNotes(page = state.currentPage + 1)
    }

    /**
     * Load custom field definitions from settings.
     */
    private fun loadCustomFields() {
        viewModelScope.launch {
            try {
                val fields = apiService.request<List<CustomFieldDefinition>>(
                    "GET",
                    "/api/settings/custom-fields",
                )
                _uiState.update {
                    it.copy(
                        customFields = fields.filter {  field ->
                            field.visibleToVolunteers && field.context == "note"
                        },
                    )
                }
            } catch (_: Exception) {
                // Custom fields are optional — silent failure is acceptable
            }
        }
    }

    /**
     * Select a note for detail view.
     */
    fun selectNote(note: DecryptedNote) {
        _uiState.update { it.copy(selectedNote = note) }
    }

    /**
     * Clear the selected note (navigating back from detail).
     */
    fun clearSelectedNote() {
        _uiState.update { it.copy(selectedNote = null) }
    }

    /**
     * Create a new encrypted note.
     *
     * @param text The note body text
     * @param fieldValues Map of custom field name -> value
     */
    fun createNote(text: String, fieldValues: Map<String, String>, conversationId: String? = null, callId: String? = null, transcript: String? = null) {
        viewModelScope.launch {
            _uiState.update { it.copy(isSaving = true, saveError = null, saveSuccess = false) }

            try {
                // Build the payload with custom field values
                val fields: Map<String, JsonElement>? = if (fieldValues.isNotEmpty()) {
                    fieldValues.mapValues { (name, value) ->
                        val fieldDef = _uiState.value.customFields.find { it.name == name }
                        when (fieldDef?.type) {
                            "checkbox" -> JsonPrimitive(value.toBooleanStrictOrNull() ?: false)
                            "number" -> {
                                val num = value.toIntOrNull()
                                if (num != null) JsonPrimitive(num) else JsonPrimitive(value)
                            }
                            else -> JsonPrimitive(value)
                        }
                    }
                } else {
                    null
                }

                // Include transcript in note text if provided
                val noteText = if (transcript != null) {
                    "$text\n\n--- Transcript ---\n$transcript"
                } else {
                    text
                }
                val payload = NotePayload(text = noteText, fields = fields)
                val payloadJson = json.encodeToString(NotePayload.serializer(), payload)

                val encrypted = cryptoService.encryptNote(payloadJson, sessionState.adminPubkeys)

                // Separate author envelope from admin envelopes.
                // The first envelope is always for the author (our pubkey).
                val authorEnv = encrypted.envelopes.first()
                val adminEnvs = encrypted.envelopes.drop(1)

                val request = CreateNoteBody(
                    encryptedContent = encrypted.ciphertext,
                    authorEnvelope = CreateNoteBodyAuthorEnvelope(
                        ephemeralPubkey = authorEnv.ephemeralPubkey,
                        wrappedKey = authorEnv.wrappedKey,
                    ),
                    adminEnvelopes = adminEnvs.map { env ->
                        CreateNoteBodyAdminEnvelope(
                            pubkey = env.recipientPubkey,
                            wrappedKey = env.wrappedKey,
                            ephemeralPubkey = env.ephemeralPubkey,
                        )
                    },
                    conversationID = conversationId,
                    callID = callId,
                )

                apiService.request<NoteResponse>("POST", apiService.hp("/api/notes"), request)

                _uiState.update { it.copy(isSaving = false, saveSuccess = true) }

                // Refresh the notes list to include the new note
                refresh()
            } catch (e: Exception) {
                _uiState.update {
                    it.copy(
                        isSaving = false,
                        saveError = e.message ?: "Failed to save note",
                    )
                }
            }
        }
    }

    /**
     * Clear the save success flag after the UI has navigated back.
     */
    fun clearSaveSuccess() {
        _uiState.update { it.copy(saveSuccess = false) }
    }

    /**
     * Clear the save error.
     */
    fun clearSaveError() {
        _uiState.update { it.copy(saveError = null) }
    }

    fun dismissError() {
        _uiState.update { it.copy(error = null) }
    }

    // ---- Search ----

    /**
     * Set the search query for filtering notes.
     */
    fun setSearchQuery(query: String) {
        _uiState.update { it.copy(searchQuery = query) }
    }

    /**
     * Get notes filtered by the current search query.
     */
    fun filteredNotes(): List<DecryptedNote> {
        val query = _uiState.value.searchQuery.lowercase()
        if (query.isBlank()) return _uiState.value.notes
        return _uiState.value.notes.filter { note ->
            note.text.lowercase().contains(query)
        }
    }

    // ---- Note Editing ----

    /**
     * Enter edit mode for the selected note.
     */
    fun startEditing() {
        _uiState.update { it.copy(isEditing = true) }
    }

    /**
     * Cancel editing and return to read mode.
     */
    fun cancelEditing() {
        _uiState.update { it.copy(isEditing = false) }
    }

    /**
     * Update an existing note with new encrypted content.
     *
     * @param noteId The note to update
     * @param text Updated plaintext note body
     * @param fieldValues Updated custom field values
     */
    fun updateNote(noteId: String, text: String, fieldValues: Map<String, String>) {
        viewModelScope.launch {
            _uiState.update { it.copy(isSaving = true, saveError = null) }

            try {
                val fields: Map<String, JsonElement>? = if (fieldValues.isNotEmpty()) {
                    fieldValues.mapValues { (name, value) ->
                        val fieldDef = _uiState.value.customFields.find { it.name == name }
                        when (fieldDef?.type) {
                            "checkbox" -> JsonPrimitive(value.toBooleanStrictOrNull() ?: false)
                            "number" -> {
                                val num = value.toIntOrNull()
                                if (num != null) JsonPrimitive(num) else JsonPrimitive(value)
                            }
                            else -> JsonPrimitive(value)
                        }
                    }
                } else {
                    null
                }

                val payload = NotePayload(text = text, fields = fields)
                val payloadJson = json.encodeToString(NotePayload.serializer(), payload)

                val encrypted = cryptoService.encryptNote(payloadJson, sessionState.adminPubkeys)

                val authorEnv = encrypted.envelopes.first()
                val adminEnvs = encrypted.envelopes.drop(1)

                val request = CreateNoteBody(
                    encryptedContent = encrypted.ciphertext,
                    authorEnvelope = CreateNoteBodyAuthorEnvelope(
                        ephemeralPubkey = authorEnv.ephemeralPubkey,
                        wrappedKey = authorEnv.wrappedKey,
                    ),
                    adminEnvelopes = adminEnvs.map { env ->
                        CreateNoteBodyAdminEnvelope(
                            pubkey = env.recipientPubkey,
                            wrappedKey = env.wrappedKey,
                            ephemeralPubkey = env.ephemeralPubkey,
                        )
                    },
                )

                apiService.request<NoteResponse>("PUT", apiService.hp("/api/notes/$noteId"), request)

                // Update local state with edited content
                val updatedNote = _uiState.value.selectedNote?.copy(
                    text = text,
                    fields = fields,
                )
                _uiState.update {
                    it.copy(
                        isSaving = false,
                        isEditing = false,
                        selectedNote = updatedNote,
                        saveSuccess = true,
                    )
                }

                refresh()
            } catch (e: Exception) {
                _uiState.update {
                    it.copy(
                        isSaving = false,
                        saveError = e.message ?: "Failed to update note",
                    )
                }
            }
        }
    }

    // ---- Thread Replies ----

    /**
     * Load replies for the currently selected note.
     */
    fun loadReplies(noteId: String) {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoadingReplies = true) }
            try {
                val response = apiService.request<NoteRepliesResponse>(
                    "GET", apiService.hp("/api/notes/$noteId/replies"),
                )
                val decrypted = response.replies.mapNotNull { reply ->
                    decryptReply(reply)
                }
                _uiState.update { it.copy(replies = decrypted, isLoadingReplies = false) }
            } catch (_: Exception) {
                _uiState.update { it.copy(isLoadingReplies = false) }
            }
        }
    }

    /**
     * Send an encrypted reply to a note thread.
     */
    fun sendReply(noteId: String, text: String) {
        if (text.isBlank()) return
        viewModelScope.launch {
            _uiState.update { it.copy(isSendingReply = true) }
            try {
                val encrypted = cryptoService.encryptNote(text, sessionState.adminPubkeys)

                val readerEnvelopes = encrypted.envelopes.map { env ->
                    CreateReplyBodyReaderEnvelope(
                        pubkey = env.recipientPubkey,
                        wrappedKey = env.wrappedKey,
                        ephemeralPubkey = env.ephemeralPubkey,
                    )
                }

                val request = CreateReplyBody(
                    encryptedContent = encrypted.ciphertext,
                    readerEnvelopes = readerEnvelopes,
                )

                apiService.requestNoContent("POST", apiService.hp("/api/notes/$noteId/replies"), request)

                // Increment local reply count
                val updatedNote = _uiState.value.selectedNote?.copy(
                    replyCount = (_uiState.value.selectedNote?.replyCount ?: 0) + 1,
                )
                _uiState.update { it.copy(isSendingReply = false, selectedNote = updatedNote) }

                // Reload replies
                loadReplies(noteId)
            } catch (e: Exception) {
                _uiState.update {
                    it.copy(
                        isSendingReply = false,
                        saveError = e.message ?: "Failed to send reply",
                    )
                }
            }
        }
    }

    /**
     * Clear replies when leaving note detail.
     */
    fun clearReplies() {
        _uiState.update { it.copy(replies = emptyList()) }
    }

    /**
     * Decrypt a single reply by finding our envelope and calling CryptoService.
     */
    private suspend fun decryptReply(reply: NoteReply): DecryptedReply? {
        val ourPubkey = cryptoService.pubkey ?: return null
        val envelope = reply.readerEnvelopes.find { it.pubkey == ourPubkey } ?: return null
        return try {
            val payload = cryptoService.decryptNote(reply.encryptedContent, envelope)
            if (payload != null) {
                DecryptedReply(
                    id = reply.id,
                    authorPubkey = reply.authorPubkey,
                    text = payload.text,
                    createdAt = reply.createdAt,
                )
            } else {
                null
            }
        } catch (_: Exception) {
            null
        }
    }

    /**
     * Decrypt a single note by finding our envelope and calling CryptoService.
     */
    private suspend fun decryptNote(note: NoteResponse): DecryptedNote? {
        val ourPubkey = cryptoService.pubkey ?: return null

        // Find the envelope for our pubkey: check authorEnvelope first (if we're
        // the author), then adminEnvelopes (if we're an admin reader).
        val envelope: RecipientEnvelope = if (note.authorPubkey == ourPubkey && note.authorEnvelope != null) {
            RecipientEnvelope(
                pubkey = ourPubkey,
                wrappedKey = note.authorEnvelope!!.wrappedKey,
                ephemeralPubkey = note.authorEnvelope!!.ephemeralPubkey,
            )
        } else {
            note.adminEnvelopes?.find { it.pubkey == ourPubkey }?.let { adminEnv ->
                RecipientEnvelope(
                    pubkey = adminEnv.pubkey,
                    wrappedKey = adminEnv.wrappedKey,
                    ephemeralPubkey = adminEnv.ephemeralPubkey,
                )
            } ?: return null
        }

        return try {
            val payload = cryptoService.decryptNote(note.encryptedContent, envelope)
            if (payload != null) {
                DecryptedNote(
                    id = note.id,
                    text = payload.text,
                    fields = payload.fields,
                    authorPubkey = note.authorPubkey,
                    callId = note.callID,
                    conversationId = note.conversationID,
                    replyCount = note.replyCount?.toInt() ?: 0,
                    createdAt = note.createdAt,
                    updatedAt = note.updatedAt,
                )
            } else {
                null
            }
        } catch (_: Exception) {
            null
        }
    }
}

/**
 * Extension to extract a display-friendly string from a [JsonElement] custom field value.
 */
fun JsonElement.displayValue(): String {
    val primitive = this.jsonPrimitive
    return when {
        primitive.booleanOrNull != null -> if (primitive.booleanOrNull == true) "Yes" else "No"
        primitive.contentOrNull != null -> primitive.content
        else -> toString()
    }
}

File: apps/ios/Sources/ViewModels/NotesViewModel.swift
import Foundation
import UIKit

// MARK: - NotesViewModel

/// View model for the Notes tab. Fetches encrypted notes from the API, decrypts them
/// using CryptoService, and manages pagination and custom field definitions.
@Observable
final class NotesViewModel {
    private let apiService: HubAPIServiceProtocol
    private let cryptoService: CryptoService

    // MARK: - Public State

    /// Decrypted notes, sorted by creation date (newest first).
    var notes: [DecryptedNote] = []

    /// Custom field definitions fetched from the server.
    var customFields: [CustomFieldDefinition] = []

    /// Whether the initial load is in progress.
    var isLoading: Bool = false

    /// Whether a page load is in progress (for pagination).
    var isLoadingMore: Bool = false

    /// Error message from the last failed operation.
    var errorMessage: String?

    /// Whether there are more notes to load.
    var hasMore: Bool = true

    /// Total number of notes on the server.
    var totalCount: Int = 0

    /// Whether the note creation sheet is shown.
    var showCreateSheet: Bool = false

    // MARK: - Private State

    private var currentPage: Int = 1
    private let pageSize: Int = 50
    private var encryptedNotes: [EncryptedNoteResponse] = []

    // MARK: - Initialization

    init(apiService: HubAPIServiceProtocol, cryptoService: CryptoService) {
        self.apiService = apiService
        self.cryptoService = cryptoService
    }

    // MARK: - Data Loading

    /// Load the first page of notes and custom field definitions.
    func loadNotes() async {
        guard !isLoading else { return }
        isLoading = true
        errorMessage = nil
        currentPage = 1

        async let notesResult: Void = fetchNotesPage(page: 1, replacing: true)
        async let fieldsResult: Void = fetchCustomFields()

        await notesResult
        await fieldsResult

        isLoading = false
    }

    /// Load the next page of notes (pagination).
    func loadMoreNotes() async {
        guard !isLoadingMore, hasMore else { return }
        isLoadingMore = true

        let nextPage = currentPage + 1
        await fetchNotesPage(page: nextPage, replacing: false)

        isLoadingMore = false
    }

    /// Refresh notes (pull-to-refresh).
    func refresh() async {
        isLoading = false
        await loadNotes()
    }

    // MARK: - Note Creation

    /// Encrypt and create a new note using HPKE envelope encryption.
    ///
    /// - Parameters:
    ///   - text: The note body text.
    ///   - fields: Custom field values keyed by field name.
    ///   - callId: Optional associated call ID.
    ///   - conversationId: Optional associated conversation ID.
    ///   - adminPubkeys: Admin encryption public keys (X25519) for envelope encryption.
    func createNote(
        text: String,
        fields: [String: AnyCodableValue]?,
        callId: String?,
        conversationId: String?,
        adminPubkeys: [String]
    ) async throws {
        let payload = NotePayload(text: text, fields: fields)
        let encoder = JSONEncoder()
        encoder.keyEncodingStrategy = .convertToSnakeCase
        let payloadJSON = String(data: try encoder.encode(payload), encoding: .utf8) ?? "{}"

        // Build full recipient list: our encryption key + admin encryption keys
        var recipientPubkeys: [String] = []
        if let ourPubkey = cryptoService.encryptionPubkeyHex {
            recipientPubkeys.append(ourPubkey)
        }
        for adminPubkey in adminPubkeys where !recipientPubkeys.contains(adminPubkey) {
            recipientPubkeys.append(adminPubkey)
        }

        let result = try cryptoService.encryptNote(payload: payloadJSON, recipientPubkeys: recipientPubkeys)

        // Map HPKE envelopes to the protocol wire format
        let authorEnvelope: ProtocolKeyEnvelope?
        let adminEnvelopes: [RecipientEnvelope]?

        if let ourPubkey = cryptoService.encryptionPubkeyHex,
           let ours = result.envelopes.first(where: { $0.pubkey == ourPubkey }) {
            authorEnvelope = ProtocolKeyEnvelope(
                ephemeralPubkey: ours.envelope.enc,
                wrappedKey: ours.envelope.ct
            )
        } else {
            authorEnvelope = nil
        }

        adminEnvelopes = result.envelopes
            .filter { $0.pubkey != cryptoService.encryptionPubkeyHex }
            .map { env in
                RecipientEnvelope(
                    ephemeralPubkey: env.envelope.enc,
                    pubkey: env.pubkey,
                    wrappedKey: env.envelope.ct
                )
            }

        let request = CreateNoteRequest(
            callId: callId,
            conversationId: conversationId,
            encryptedContent: result.ciphertextHex,
            authorEnvelope: authorEnvelope,
            adminEnvelopes: adminEnvelopes
        )

        let _: EncryptedNoteResponse = try await apiService.request(
            method: "POST",
            path: apiService.hp("/api/notes"),
            body: request
        )

        // Haptic feedback on success
        let generator = UIImpactFeedbackGenerator(style: .medium)
        generator.impactOccurred()

        // Reload notes to include the new one
        await refresh()
    }

    // MARK: - Note Decryption

    /// Find the matching envelope for our pubkey and decrypt the note using HPKE.
    func decryptNote(_ encrypted: EncryptedNoteResponse) -> DecryptedNote? {
        guard let ourPubkey = cryptoService.encryptionPubkeyHex else { return nil }

        // Find our envelope — check author envelope first (volunteer's own note)
        var envelope: HpkeEnvelope?

        if encrypted.authorPubkey == ourPubkey, let authorEnv = encrypted.authorEnvelope {
            envelope = HpkeEnvelope(
                v: 3,
                labelId: 0,
                enc: authorEnv.ephemeralPubkey,
                ct: authorEnv.wrappedKey
            )
        }

        // Then check admin envelopes
        if envelope == nil, let adminEnvs = encrypted.adminEnvelopes {
            if let ourEnvelope = adminEnvs.first(where: { $0.pubkey == ourPubkey }) {
                envelope = HpkeEnvelope(
                    v: 3,
                    labelId: 0,
                    enc: ourEnvelope.ephemeralPubkey,
                    ct: ourEnvelope.wrappedKey
                )
            }
        }

        guard let hpkeEnvelope = envelope else {
            return nil
        }

        do {
            let decryptedJSON = try cryptoService.decryptNote(
                ciphertextHex: encrypted.encryptedContent,
                envelope: hpkeEnvelope
            )

            let decoder = JSONDecoder()
            decoder.keyDecodingStrategy = .convertFromSnakeCase
            let payload = try decoder.decode(NotePayload.self, from: Data(decryptedJSON.utf8))

            return DecryptedNote(
                id: encrypted.id,
                payload: payload,
                authorPubkey: encrypted.authorPubkey,
                callId: encrypted.callID,
                conversationId: encrypted.conversationID,
                createdAt: DateFormatting.parseISO(encrypted.createdAt) ?? Date(),
                updatedAt: DateFormatting.parseISO(encrypted.updatedAt)
            )
        } catch {
            return nil
        }
    }

    // MARK: - Private Helpers

    private func fetchNotesPage(page: Int, replacing: Bool) async {
        do {
            let response: NotesListResponse = try await apiService.request(
                method: "GET",
                path: apiService.hp("/api/notes?page=\(page)&limit=\(pageSize)")
            )

            if replacing {
                encryptedNotes = response.notes
            } else {
                encryptedNotes.append(contentsOf: response.notes)
            }

            totalCount = response.total
            currentPage = page
            hasMore = encryptedNotes.count < response.total

            // Decrypt all notes
            notes = encryptedNotes.compactMap(decryptNote)
        } catch {
            if case APIError.noBaseURL = error {
                // Hub not configured — show empty state, no error
            } else {
                errorMessage = error.localizedDescription
            }
        }
    }

    private func fetchCustomFields() async {
        do {
            let response: CustomFieldsResponse = try await apiService.request(
                method: "GET",
                path: apiService.hp("/api/settings/custom-fields")
            )
            customFields = response.fields
                .filter { $0.visibleToVolunteers }
                .sorted { $0.order < $1.order }
        } catch {
            // Custom fields are optional — silently continue without them
            customFields = []
        }
    }

}
